### PR TITLE
Limiting euler function to positive numbers

### LINKED ISSUE #8786: Euler number of negative integers
Currently `euler(-1)`, `euler(-3)`, etc. all evaluate to 0, as the first test in `euler.eval` is

``` python
        if m.is_odd:
            return S.Zero
```

but euler numbers are not defined on negative numbers, so this is wrong (and should probably raise a `ValueError`).

### PR DESCRIPTION
**Pull request by sahilshekhawat**
Raising valueerror on negative numbers
Fixes #8786

### CODE REVIEW

**Comment by pelegm** [comment]
Please add relevant tests.

File: sympy/functions/combinatorial/numbers.py [hunk 1]
```diff
@@ -705,13 +705,16 @@ class euler(Function):
...
     @classmethod
     def eval(cls, m):
-        if m.is_odd:
-            return S.Zero
-        if m.is_Integer and m.is_nonnegative:
-            from mpmath import mp
-            m = m._to_mpmath(mp.prec)
-            res = mp.eulernum(m, exact=True)
-            return Integer(res)
+        if not m.is_negative:
```
Comment (aktech) [inline sympy/functions/combinatorial/numbers.py]: It would be better to use `if m.is_nonnegative`.

**Comment by aktech** [comment]
Please squash commits (using git rebase).

**Comment by sahilshekhawat** [comment]
can we merge this now?

**Comment by sahilshekhawat** [comment]
@pelegm can we merge this now?

**Comment by sahilshekhawat** [comment]
@smichr can you please review this?

**Comment by sahilshekhawat** [comment]
@pelegm I have replaced the tests with ``raises`` from pytest.

**Comment by sahilshekhawat** [comment]
@asmeurer This needs to be reviewed, I guess its fit to be merged.

**Comment by pelegm** [comment]
This is not good. For example:
```python
>>> n = Symbol('n', integer=True, odd=True)
>>> euler(n)
0
```
(should be `euler(n)` instead, as we don't know `n` is nonnegative).

**Comment by sahilshekhawat** [comment]
This PR only puts a restriction on the negative numbers in ``euler()`` and this is, I feel, is a separate issue and can be addressed separately.

**Comment by pelegm** [comment]
It might be in a different PR, I agree, but if you're making changes already, I'd solve easy critical bugs while doing so... It is your decision, I just post here my opinion.

**Comment by sahilshekhawat** [comment]
I really appreciate that you are helping me in reviewing this PR and I respect your views. Its true that I can fix that right here but I would like to raise a separate PR to address this issue, just to address separate issues separately. Does that make any sense? Although, I can fix this in this PR too but that will only delay its merge.

**Comment by sahilshekhawat** [comment]
Also, ``euler()`` return 0 for any odd number. This seems correct to me. It should not return ``euler(n)`` when n is odd.

**Comment by pelegm** [comment]
> Its true that I can fix that right here but I would like to raise a separate PR to address this issue, just to address separate issues separately. Does that make any sense?

Well, it is true indeed that these may be thought of as two different subjects. However, fixing the latter will require a rewrite of that function anyway, so it seems like double work. For example, if I would try to fix the latter, I'd probably fix the first on the fly, since otherwise someone else would have to rewrite my code anyway.

> Seems correct to me.

I understand your feeling about that, and if I remember correctly I thought the same a few months ago. However, there seems to be a strong convention here about questions of this type. See for example in #8490.

I think the main idea is that `function(n)` will not return a number unless every substitution of a number instead of `n` for which the assumptions on `n` hold will yield that very same number. This is not the case here.

**Comment by oscargus** [comment]
This functionality seem to have been implemented elsewhere since, so I am closing this.

**Pull request closed without merging**